### PR TITLE
[Runtime] Change default alignment to 64 bytes.

### PR DESCRIPTION
**Pull request by jwfromm**
One change made in #5252 (which added support for Hexagon to the runtime) was increasing the byte alignment from 64 to 128. This can cause problems when interacting with `dlpack`. For example `tests/python/contrib/test_dlpack.py` has a high chance of failing when run locally due to torch returning tensors with 64 rather than 128 byte alignment. I'm not sure why it doesnt fail in CI, perhaps the consistency of the environment always returns an appropriately aligned tensor.

Changing the default alignment allows interoperability with both torch and newer versions of numpy that support dlpack. I've slightly modified the torch test to run multiple times to make sure its behavior is consistent.

@kparzysz-quic do you think there are any drawbacks to decreasing the bit alignment requirement?

cc @areusch

### CODE REVIEW

**Comment by jwfromm** [comment]
@tqchen might be worth having you take a look to see if there are any potential problems with this change.

File: python/tvm/tir/schedule/schedule.py [hunk 1]
```diff
@@ -2092,9 +2092,9 @@ def before_tensorize(
 
             @T.prim_func
             def mma_desc(a: T.handle, b: T.handle, c: T.handle) -> None:
-                A = T.match_buffer(a, (16, 16), align=128, offset_factor=1)
-                B = T.match_buffer(b, (16, 16), align=128, offset_factor=1)
-                C = T.match_buffer(c, (16, 16), align=128, offset_factor=1)
+                A = T.match_buffer(a, (16, 16), align=64, offset_factor=1)
+                B = T.match_buffer(b, (16, 16), align=64, offset_factor=1)
+                C = T.match_buffer(c, (16, 16), align=64, offset_factor=1)
```
Comment (junrushao) [inline python/tvm/tir/schedule/schedule.py:2097]: mma does use `align=128`, so let's revert this change here :-)

Comment (jwfromm) [reply python/tvm/tir/schedule/schedule.py]: Ah make sense, reverted. Is there anywhere else we'd want 128 byte alignment? I just did a global search and replace.

Comment (junrushao) [reply python/tvm/tir/schedule/schedule.py]: not super familiar with non-cuda cases. probably @Hzfengsy @spectrometerHBH could help!

Comment (Hzfengsy) [reply python/tvm/tir/schedule/schedule.py]: Not sure about other backends. But it's fine if we can pass the tests

File: python/tvm/tir/tensor_intrin/cuda.py [hunk 2]
```diff
@@ -699,18 +697,18 @@ def wmma_sync_desc(a: T.handle, b: T.handle, c: T.handle) -> None:
...
             maybe_swap(k_dim, n_dim),
             in_dtype,
-            align=128,
+            align=64,
             offset_factor=16,
             scope="wmma.matrix_b",
         )
         C = T.match_buffer(
-            c, (m_dim, n_dim), out_dtype, align=128, offset_factor=16, scope="wmma.accumulator"
+            c, (m_dim, n_dim), out_dtype, align=64, offset_factor=16, scope="wmma.accumulator"
```
Comment (masahi) [inline python/tvm/tir/tensor_intrin/cuda.py:711]: I think all CUDA code wants 128 bit alignment in practice. cc @vinx13

**Comment by u99127** [comment]
> One change made in #5252 (which added support for Hexagon to the runtime) was increasing the byte alignment from 64 to 128. This can cause problems when interacting with `dlpack`. For example `tests/python/contrib/test_dlpack.py` has a high chance of failing when run locally due to torch returning tensors with 64 rather than 128 byte alignment. I'm not sure why it doesnt fail in CI, perhaps the consistency 



Nit: Bits and bytes conflict in commit message vs title. Is it 128byte alignment or 128bit alignment 😄 ...

**Comment by kparzysz-quic** [comment]
Hexagon HVX vectors should be aligned at 128-byte boundary.  If not, we'd have to use unaligned loads/stores.  I think we have some code that relies on the current alignment.

This should really be a target option, maybe it's time to refactor this code...

File: python/tvm/tir/tensor_intrin/cuda.py (hunk 2 above)
Comment (jwfromm) [reply python/tvm/tir/tensor_intrin/cuda.py:711]: Yeah I'll update this spot as well.

**Comment by kparzysz-quic** [comment]
> Fortunately Hexagon does use its own target specific alignment `kHexagonAllocAlignment`, so it would not be affected by this change.

I guess it's ok then.

**Comment by tqchen** [comment]
There are two parts of the perspective that comes into the story:

- P0: The alignment of the allocator
- P1: The alignment assumption of the compiler

P1 affects how compiler generates code, e.g. if we assume that the code is aligned to 128bytes, then compiler can generate aligned store/load safely. P0 affects the array we allocate, we always need to allocate memory that is larger than P1 assumption made by the compiler.

Right now the two are roughly tied together, with the exception of `kHexagonAllocAlignment`, which enforces P0. 

The PyTorch issue was caused by P1 -- we are making assumption of 128 byte alignment while pytorch arrays are commonly aligned to 64 bytes.

Currently, P1 is supplied by the buffer construct, which do not (yet) depend on the target.  Agree with @kparzysz-quic that minimum alignment requirement can be part of the target, and we need to find good ways to set them onto the buffer. 

As a short workaround, we can detect P1 settings(of buffers) in early stage of TIR compilation and amend them for devices like HVX

**Comment by jwfromm** [comment]
Thanks for that context @tqchen. Would you like to see any of your suggested improvements in this PR or do you think this is an appropriate first step?

**Comment by ekalda** [comment]
I don't think this will cause problems for Ethos-U since its accesses are 16 byte aligned (but let's see if the tests pass). Regarding to why this failure is not visible in CI, could it be related to this issue? https://github.com/apache/tvm/issues/12529

**Comment by tqchen** [comment]
The previous comment https://github.com/apache/tvm/pull/3163#discussion_r283733488 would make it work for temp data allocations(change https://github.com/apache/tvm/blob/8146a9bf2c9288103c4af834a5f14f13b50aa3c8/src/tir/transforms/ir_utils.h#L166 to depend on Target and return different value in different setting).

For buffer declarations, the main issue is that Buffer creation is not target dependent as of now, as a result, https://github.com/apache/tvm/blob/main/src/tir/ir/buffer.cc#L570 defaults to the global valu.

We can introduce a minimum_data_alignment  attr to target, and modify buffer->data_alignment when the user run things through build.

**Comment by jwfromm** [comment]
It seems like we would need to do a significant rework to use target specific alignment for this to work. Some of the new tensorization intrinsics allocate buffers with the default alignment that now conflict with explicit 128 byte alignment. I'm going to close this PR as I dont think I have the expertise to do the refactorization at this time. We should just note that in the meantime dlpack tensors will not be able to be passed to TVM consistently.

**Comment by jwfromm** [comment]
After chatting with @vinx13, 64 byte alignment should actually be fine for most of cuda. I'm reponing this PR as I think its still a positive change while we work out a more permanent solution.
